File: test/four/app/src/main/java/com/z3d/md5/MeshImportSort.java
package com.z3d.md5;

import com.z3d.md5.vo.Md5MeshData;
import com.z3d.md5.vo.ObjectBone;
import com.z3d.md5.vo.ObjectTri;
import com.z3d.md5.vo.ObjectUv;
import com.z3d.md5.vo.ObjectWeight;

import java.util.ArrayList;
import java.util.List;

public class MeshImportSort {
    public void processMesh(Md5MeshData meshData) {

        List<ObjectWeight> weightAry = new ArrayList<>();
        int i = 0;
        for (i = 0; i < meshData.weightItem.size(); i++) {
            weightAry.add(meshData.weightItem.get(i).clone());
        }
        List<Integer> mapkeyAry  = this.getMapValue(meshData.boneItem);
        for (i = 0; i < weightAry.size(); i++) {
            weightAry.get(i).boneId = mapkeyAry.get(weightAry.get(i).boneId);
        }
        meshData.weightItem = weightAry;
        this.processForAgal(meshData);
    }

    private void processForAgal(Md5MeshData meshData) {
        int beginKey = 1;

        List<ObjectUv> uvItem = meshData.uvItem;
        List<ObjectWeight> weightItem = meshData.weightItem;
        List<ObjectTri> triItem = meshData.triItem;

        List<Float> uvArray  = new ArrayList<>();
        List<List<Float>> ary  = new ArrayList<>();
        ary.add(new ArrayList<>());
        ary.add(new ArrayList<>());
        ary.add(new ArrayList<>());
        ary.add(new ArrayList<>());

        List<Float> boneWeightAry = new ArrayList<>();
        List<Integer> bonetIDAry = new ArrayList<>();
        List<Integer> indexAry  = new ArrayList<>();

        int skipNum;
        int beginIndex;
        int allNum;

        List<Integer> boneUseAry  = new ArrayList<>();

        for (int i = 0; i < uvItem.size(); i++) {
            beginIndex = uvItem.get(i).a;
            allNum = uvItem.get(i).b;
            for (skipNum = 0; skipNum < 4; skipNum++) {
                if (skipNum < allNum) {
//                    boneUseAry.push((weightItem[beginIndex + skipNum].boneId));
                    boneUseAry.add(weightItem.get(beginIndex + skipNum).boneId);
                } else {
//                    boneUseAry.push(boneUseAry[0]);
                    boneUseAry.add(boneUseAry.get(0));
                }
            }
        }
        boneUseAry = this.getboneNum(boneUseAry);

        for (int i = 0; i < uvItem.size(); i++) {
            beginIndex = uvItem.get(i).a;
            allNum = uvItem.get(i).b;
            for (skipNum = 0; skipNum < 4; skipNum++) {
                if (skipNum < allNum) {
                    ary.get(skipNum).add(weightItem.get(beginIndex + skipNum).x);
                    ary.get(skipNum).add(weightItem.get(beginIndex + skipNum).y);
                    ary.get(skipNum).add(weightItem.get(beginIndex + skipNum).z);
                    bonetIDAry.add(boneUseAry.indexOf((weightItem.get(beginIndex + skipNum).boneId)));
                    boneWeightAry.add(weightItem.get(beginIndex + skipNum).w);

                } else {

                    ary.get(skipNum).add(0f);
                    ary.get(skipNum).add(0f);
                    ary.get(skipNum).add(0f);
                    bonetIDAry.add(boneUseAry.indexOf(0));
                    boneWeightAry.add(0.0f);
                }
            }

            uvArray.add(uvItem.get(i).x);
            uvArray.add(uvItem.get(i).y);
        }
        meshData.boneNewIDAry = getShortArrByInter(boneUseAry);

        for (int i = 0; i < triItem.size(); i++) {
            indexAry.add(triItem.get(i).t0);
            indexAry.add(triItem.get(i).t1);
            indexAry.add(triItem.get(i).t2);
        }
        meshData.faceNum = indexAry.size() / 3;
        meshData.treNum = indexAry.size();
        this.uplodToGpu(meshData, uvArray, ary, boneWeightAry, bonetIDAry, getShortArrByInter(indexAry));
    }

    private void uplodToGpu(Md5MeshData meshData, List<Float> uvArray, List<List<Float>> ary, List<Float> boneWeightAry, List<Integer> bonetIDAry, List<Short> indexAry) {
        meshData.boneWeightAry = boneWeightAry;
        meshData.boneIDAry  = new ArrayList<>();
        for (int i = 0; i < bonetIDAry.size(); i++) {
            meshData.boneIDAry .add(Math.max(bonetIDAry.get(i), 0f));
        }
    }


    private List<Short> getShortArrByInter(List<Integer> arr){
        List<Short> outArr=new ArrayList<>();
        for(int i=0;i<arr.size();i++){
            outArr.add(arr.get(i).shortValue());
        }
        return  outArr;
    }

    private List<Integer> getboneNum(List<Integer> ary) {
        List<Integer> numAry= new ArrayList<>();
        for (int i = 0; i < ary.size(); i++) {
            if (numAry.indexOf(ary.get(i)) == -1) {
                numAry.add(ary.get(i));
            }
        }

        return numAry;
    }

    private List<Integer> getMapValue(List<ObjectBone> targetAry) {
        List<ObjectBone> newTargetAry = MeshToObjUtils.getStorNewTargerArr(targetAry);
        List<Integer> mapkeyAry = new ArrayList<>();//新旧ID映射关系
        for (int i = 0; i < targetAry.size(); i++) {
            int index = newTargetAry.indexOf(targetAry.get(i));
            mapkeyAry.add(index);
        }
        return mapkeyAry;
    }
}

File: metalios/listios/StorageStarted/pan3d/md5/MeshImportSort.m
//
//  MeshImportSort.m
//  iosgl
//
//  Created by pan3dme on 2020/11/12.
//  Copyright © 2020 zhao. All rights reserved.
//

#import "MeshImportSort.h"
#import "ObjectWeight.h"
#import "MeshToObjUtils.h"
#import "ObjectUv.h"
#import "ObjectWeight.h"
#import "MeshData.h"
#import "ObjectTri.h"

@implementation MeshImportSort
-(void)processMesh:(Md5MeshData*)meshData
{
    
    NSMutableArray* weightAry  = [[NSMutableArray alloc]init];
    for (int i = 0; i < meshData.weightItem.count; i++) {
        [weightAry addObject:     [((ObjectWeight*)(meshData.weightItem[i])) clone]];
    }
    NSArray* mapkeyAry  = [self getMapValue:meshData.boneItem];
    for (int i = 0; i < weightAry.count; i++) {
        int boneId=[mapkeyAry[((ObjectWeight*)(weightAry[i])).boneId] intValue];
        ((ObjectWeight*) weightAry[i]).boneId=[mapkeyAry[boneId] intValue];
    }
    meshData.weightItem = weightAry;
    [self processForAgal:meshData];
    
}
-(int)indexOf:(NSArray*)arr find:(NSObject*)find
{
    NSInteger idx=[arr indexOfObject:find];
    if(idx>arr.count){
        return -1;
    }else{
        return (int)idx;
    }
}
-(void)processForAgal:(Md5MeshData*)meshData
{
    int beginKey = 1;
    NSArray* uvItem= meshData.uvItem;//: Array<ObjectUv>
    NSArray* weightItem = meshData.weightItem;// Array<ObjectWeight>
    NSArray* triItem = meshData.triItem; //Array<ObjectTri>
    NSMutableArray* uvArray  = [[NSMutableArray alloc]init];
//    NSMutableArray* ary= [[NSMutableArray alloc]init];//Array<Array<number>>
    
    NSArray *ary = @[[[NSMutableArray alloc]init], [[NSMutableArray alloc]init], [[NSMutableArray alloc]init], [[NSMutableArray alloc]init]];
    
    
    NSMutableArray* boneWeightAry=[[NSMutableArray alloc]init];
    NSMutableArray* bonetIDAry=[[NSMutableArray alloc]init];
    NSMutableArray* indexAry=[[NSMutableArray alloc]init];
    
    int skipNum;
    int beginIndex;
    int allNum;
    
    NSMutableArray* boneUseAry = [[NSMutableArray alloc]init];
    
    for (int i = 0; i < uvItem.count; i++) {
        beginIndex = ((ObjectUv*)(uvItem[i])).a;
        allNum = ((ObjectUv*)(uvItem[i])).b;
        for (skipNum = 0; skipNum < 4; skipNum++) {
            if (skipNum < allNum) {
                int boneId=((ObjectWeight*)(weightItem[beginIndex + skipNum])).boneId;
                [boneUseAry addObject: [NSNumber numberWithInt:boneId]];
            } else {
                [boneUseAry addObject:boneUseAry[0]];
            }
        }
    }
    boneUseAry = [self getboneNum:boneUseAry];
    for (int i = 0; i < uvItem.count; i++) {
        ObjectUv* objectUv=   uvItem[i];
        beginIndex = objectUv.a;
        allNum =objectUv.b;
        for (skipNum = 0; skipNum < 4; skipNum++) {
            if (skipNum < allNum) {
                ObjectWeight* objectWeight=weightItem[beginIndex + skipNum];
                [ary[skipNum] addObject:[NSNumber numberWithFloat:objectWeight.x ]];
                [ary[skipNum] addObject:[NSNumber numberWithFloat:objectWeight.y ]];
                [ary[skipNum] addObject:[NSNumber numberWithFloat:objectWeight.z ]];
                int idx=[self indexOf: boneUseAry find:[NSNumber numberWithInteger:objectWeight.boneId]];
                [bonetIDAry addObject:[NSNumber numberWithInt:idx]];
                [boneWeightAry addObject:[NSNumber numberWithFloat:objectWeight.w]];
                
            } else {
                [ary[skipNum] addObject:[NSNumber numberWithFloat:0]];
                [ary[skipNum] addObject:[NSNumber numberWithFloat:0]];
                [ary[skipNum] addObject:[NSNumber numberWithFloat:0]];
                int idx=[self indexOf: boneUseAry find:[NSNumber numberWithInteger:0]];
                [bonetIDAry addObject:[NSNumber numberWithInt: idx]];
                [boneWeightAry addObject:[NSNumber numberWithFloat:0]];
            }
        }
        [uvArray addObject:[NSNumber numberWithFloat:objectUv.x]];
        [uvArray addObject:[NSNumber numberWithFloat:objectUv.y]];
    
    }
    meshData.boneNewIDAry = boneUseAry;
    
    for (int i = 0; i < triItem.count; i++) {
        ObjectTri* objectTri =  triItem[i];
 
        [indexAry addObject:[NSNumber numberWithInt:objectTri.t0]];
        [indexAry addObject:[NSNumber numberWithInt:objectTri.t1]];
        [indexAry addObject:[NSNumber numberWithInt:objectTri.t2]];
        
    }
    meshData.faceNum = (int)(indexAry.count / 3);
    meshData.trinum = (int)indexAry.count;
    [self uplodToGpu:meshData uvArray:uvArray ary:ary boneWeightAry:boneWeightAry bonetIDAry:bonetIDAry indexAry:indexAry];
    
 
    
}
-(void)uplodToGpu:(MeshData*)meshData uvArray:(NSArray*)uvArray  ary:(NSArray*)ary  boneWeightAry:(NSArray*)boneWeightAry  bonetIDAry:(NSArray*)bonetIDAry indexAry:(NSArray*)indexAry
{
    meshData.boneWeightAry = boneWeightAry;
    NSMutableArray* arrA   = [[NSMutableArray alloc] init];
    for (int i = 0; i < bonetIDAry.count; i++) {
        [arrA addObject:[NSNumber numberWithInt:MAX([bonetIDAry[i] intValue], 0)] ];
    }
    meshData.boneIDAry = arrA;
 
}
 
-(NSMutableArray*)getboneNum:(NSArray*)ary
{
    NSMutableArray* numAry = [[NSMutableArray alloc]init];
    for (int i = 0; i < ary.count; i++) {
        if ([numAry indexOfObject:ary[i]]>numAry.count) {
            [numAry addObject:ary[i]];
        }
    }
    return numAry;
}

-(NSArray*)getMapValue:(NSArray* )targetAry{
    
    NSArray* newTargetAry=[self.scene3D.meshToObjUtils getStorNewTargerArr: targetAry ];
    NSMutableArray* mapkeyAry  = [[NSMutableArray alloc]init ];//新旧ID映射关系
    for (int i = 0; i < targetAry.count; i++) {
        NSUInteger index= [newTargetAry indexOfObject: targetAry[i]];
        [mapkeyAry addObject:[NSNumber numberWithInteger:index]];
    }
    return mapkeyAry;
    
}

@end
